### DISCUSSION TITLE
How to get grounded predicates and functions

### DISCUSSION THREAD

**Discussion by MFaisalZaki** [JuliaPlanners/PDDL.jl#22] [Q&A] [answered]
I'm using SMT to solve the planning task. to do so I need to the groudned predicates and functions to encode them into a SMt formula but there is no clear way on how to obtain such informaion.

I have tried the following steps but it returns the lifted predicate/function:
```
domain = load_domain(joinpath("domain.pddl"));
task     = load_problem(joinpath("pfile1.pddl"));

grounded_problem = ground(domain, task);

for ele in grounded_problem.source.predicates
    println(ele)
end
```

I get a `Pair{Symbol, Main.PDDL.Signature}` which does not contain any grounded predicates.

I hope if you can help me with this.

**Comment by ztangent** [comment]
It's undocumented, but you use this function to get an iterator over all possible arguments for a predicate / fluent!

https://github.com/JuliaPlanners/PDDL.jl/blob/88644181e452c7108a5eaee61d6039c2b6bc044c/src/grounding/utils.jl#L3-L13

This doesn't do any relevance checking though. If you want the list of all "planning-relevant-ground-predicates", you can construct a GraphPlan-style planning graph using this function, and then access the `conditions` field.:

https://github.com/JuliaPlanners/SymbolicPlanners.jl/blob/c2c83e06e826dae3fae53a7e257d0df158a8d0c0/src/heuristics/pgraph.jl#L17-L27

**Comment by ztangent** [comment]
Hopefully this resolved your issue! Moving this to the Q&A section of the Discussions page.

**Reply by MFaisalZaki** [reply]
Thanks @ztangent. Yup, this works for the actions. However, I'm trying to get all boolean variables' initial values using the `initstate`, but I receive only the variables which are true. How can I get the ones that are false also?

Here is the code I used:
```
using PDDL, PlanningDomains
problemdomain   = load_domain(IPCInstancesRepo, "ipc-2002-rovers-numeric-automatic");
probleminstance = load_problem(IPCInstancesRepo, "ipc-2002-rovers-numeric-automatic", 1);

state = initstate(problemdomain, probleminstance);
for var in state.facts
  println(var)
end
```
Thanks.

**Reply by ztangent** [reply, marked as answer]
I've avoided storing all the initial values in `state.facts` because that's an inefficient representation.

However, for states produced [by compilation](https://juliaplanners.github.io/PDDL.jl/stable/tutorials/speeding_up/), you can call `PDDL.get_fluents(state)` and it will list all fluents and their associated values.

**Reply by MFaisalZaki** [reply]
thanks @ztangent for highlighting this. However, I'm getting a stack overflow when invoking `PDDL.get_fluents` for the compiled state.
Here is the function I used
```
function encodeInitialState!(problemdomain::GenericDomain, probleminstance::GenericProblem)
    c_domain, _c_state = compiled(problemdomain, probleminstance)
    _fluents = PDDL.get_fluents(_c_state)
    for (k, v) in fluents
        if isa(v, Float64)
            # do sutff
        elseif isa(v, Int64)
            # do stuff
        elseif isa(v, Bool)
            # do stuff
        else
            @assert false
        end
    end
end
```

Any ideas? :)

**Reply by ztangent** [reply]
Hmm, I'm not sure why, but one issue that can arise is that you can't actually call `compiled` from within another function due to existing limitations with how I implemented the compiler.

Given that you're trying to do this within a function, I would recommend iterating over all fluent names, and calling `groundargs` on them:

https://github.com/JuliaPlanners/PDDL.jl/blob/88644181e452c7108a5eaee61d6039c2b6bc044c/src/grounding/utils.jl#L3-L13


So you could do something like:

```julia
function groundfluents(domain::Domain, state::State, fluent::Symbol)
      return [Compound(fluent, collect(Term, args)) for args in 
                  PDDL.groundargs(domain, state, fluent)]
end

function groundfluents(domain::Domain, state::State)
       fluent_terms = Term[]
       for (name, signature) in PDDL.get_fluents(domain)
              append!(fluent_terms, groundfluents(domain, state, name)
       end
       return fluent_terms
end
```

`groundfluents` will return a list of all ground fluent terms for a state, but not their values. To get their values, you can do something like:

```julia
for fluent in groundfluents(domain, state)
       val = state[fluent]
       # Do stuff
end
```

Hope this helps!
